counsel.el

```diff
@@ -4459,12 +4463,15 @@ Return closure suitable for `ivy-format-function'."
 You can insert or kill the name or hexadecimal RGB value of the
 selected color."
   (interactive)
-  (let* ((colors (mapcar (lambda (cell)
-                           (let ((name (car cell)))
-                             (propertize name
-                                         'hex (counsel-colors--name-to-hex name)
-                                         'dups (cdr cell))))
-                         (list-colors-duplicates)))
+  (let* ((colors
+          (delete nil
```

Comment:
Better to use `delq` here.

Reply:
yes, in fact it is probably better not to allocate two lists and use mapcan instead.

Reply:
It's not always clear when it's better to use `(delq nil (mapcar ...))` vs `mapcan`. The thing about `mapcan` is it was only added in Emacs 25, so you have to use the wrapper `cl-mapcan`, and you still need to allocate a cons cell in the loop for each element you want to keep:

```el
(cl-mapcan (lambda (cell)
             (let* ((name (car cell))
                    (dups (cdr cell))
                    (hex (counsel-colors--name-to-hex name)))
               (and hex (list (propertize name 'hex hex 'dups dups)))))
           (list-colors-duplicates))
```

Benchmarking the two approaches suggests `delq`+`mapcar` is a hair faster and smaller than `cl-mapcan`. I really don't mind either way, I'm just saying there are tradeoffs, so it basically comes down to personal preference.

Reply:
Thanks for running the experimentation. 

Note that contrary to what I said, delete is destructive and does not make a new list (but remove does).

Reply:
> Note that contrary to what I said, delete is destructive and does not make a new list (but remove does).

Yup. The difference between `delq` and `delete` is that the former only accepts lists and uses `eq` for equality, thus making it the standard function for use in `(delq nil (mapcar ...))`. Using `delete` is less common and reserved for cases where equivalence under `equal` is needed.